service/apis/qsl_service.go

```diff
@@ -0,0 +1,379 @@
+				returnlist = append(returnlist, strings.Repeat("\t", len(fieldlist)-i+tabs)+"}")
+			}
+			return returnlist, nil
+		}
+		return nil, errors.New("Fields may be a string of * indicating how many levels, or a list of fields @field1,@field2,... not both [" + fieldlist + "]")
+
+	}
+
+	splitlist := strings.Split(fieldlist, ",")
+	returnlist := []string{}
+
+	for i, item := range splitlist {
+		_ = i
+		if strings.HasPrefix(item, "@") && len(item) > 1 {
+			if IsAlpha(item[1:]) {
+				returnlist = append(returnlist, strings.Repeat("\t", tabs+1)+item[1:])
+			} else {
+				return nil, errors.New("Field names must be composed of only alphanumeric characters [" + item[1:] + "]")
+			}
+
+		} else {
+			return nil, errors.New("Field names must be prefixed with @ sign and followed by an alphanumeric field name [" + item + "]")
+		}
+
+	}
+
+	returnlist = append(returnlist, strings.Repeat("\t", tabs+1)+"uid")
+	log.Debugf("fields %#v\n", returnlist)
+	return returnlist, nil
+
+}
+
+// CreateDgraphQueryHelper creates the inner nested clauses searching for relationships
+func (qa *QSLService) CreateDgraphQueryHelper(query []string, tabs int, parent string) ([]string, error) {
+
+	// string that we're going to replace
+	basequery := []string{
+		strings.Repeat("\t", tabs) + "$RELATION @filter(eq(objtype, $OBJTYPE) and$FILTERSFUNC){",
+	}
+	_ = basequery
+
+	// regex to match the string pattern
+	//r := regexp.MustCompile(`([a-zA-Z]+)\[(\@[\,\@\"\=a-zA-Z0-9\-\.\|]*)\]\{([\*|[\,\@\"\=a-zA-Z0-9\-]*)`)
+	//r := regexp.MustCompile(`([a-zA-Z]+)\[(?:(\@[\,\@\"\=a-zA-Z0-9\-\.\|\:_]*|\*))\]\{([\*|[\,\@\"\=a-zA-Z0-9\-]*)`)
+	r := regexp.MustCompile(blockRegex)
+	matches := r.FindStringSubmatch(query[0])
+	// log.Infof("%#v\n", splitquery)
+	log.Debugf("helpermatches %#v\n", matches)
+
+	// extract the values of the form objtype[filters]fields and assign to individual variables
+	objtype := strings.Title(matches[1])
+	filters := matches[2]
+	fields := matches[3]
+
+	// get a list of the metadata fields for this object type
+	metafieldslist, err := qa.metaSvc.GetMetadataFields(objtype)
+	if err != nil {
+		log.Error("err in getting metadata fields for " + objtype)
+		log.Error(err)
+		return nil, errors.New("Failed to connect to dgraph to get metadata")
+	}
+
+	if len(metafieldslist) == 0 {
+		log.Error("metadata for " + objtype + " not found in db. Will not be able to use * or find relationships")
+	}
+
+	log.Debugf("metadata fields for %s: %#v", objtype, metafieldslist)
+
+	// declare relation variable
+	relation := ""
+	// see if we can find the reverse relation from this object to its parent
+	found := false
+	// look in the list of fields for the metadata and
+	// find if there's a relationship between the parent's and this object's type
+	// e.g. if we had cluster[...]{...}.pod[...]{...} parent=cluster
+	// and we will find the pods relation to cluster is called ~cluster
+	for i, item := range metafieldslist {
+		_ = i
```

Comment:
should be able to use same small tweak here to collapse lines 216 and 217 as mentioned in other comment

also seen a few other places in PR, but I won't call each one yet